wasm/wasm_source/src/vp_user.rs

```diff
@@ -158,35 +136,200 @@ fn validate_tx(
+        }
+        false
+    };
+
+    // Metadata changes must be signed by the validator whose
+    // metadata is manipulated
+    let is_valid_metadata_change = || {
+        let metadata = storage::is_validator_metadata_key(key);
+        match metadata {
+            Some(address) => address == owner && **valid_sig,
+            None => false,
+        }
+    };
+
+    // Changes in validator state
+    let is_valid_state_change = || {
+        let state_change = storage::is_validator_state_key(key);
+        let is_valid_state = match state_change {
+            Some((address, epoch)) => {
+                let params_pre = proof_of_stake::read_pos_params(&ctx.pre())?;
+                let state_pre = proof_of_stake::validator_state_handle(address)
+                    .get(&ctx.pre(), epoch, &params_pre)?;
+
+                let params_post = proof_of_stake::read_pos_params(&ctx.post())?;
+                let state_post = proof_of_stake::validator_state_handle(
+                    address,
+                )
+                .get(&ctx.post(), epoch, &params_post)?;
+
+                match (state_pre, state_post) {
+                    (Some(pre), Some(post)) => {
+                        use proof_of_stake::types::ValidatorState::*;
+
+                        if (
+                            // Deactivation case
+                            matches!(
+                                    pre,
+                                    Consensus | BelowCapacity | BelowThreshold
+                                ) && post == Inactive)
+                            // Reactivation case
+                            || pre == Inactive && post != Inactive
+                            // Unjail case
+                            || pre == Jailed
+                                && matches!(
+                                    post,
+                                    Consensus
+                                        | BelowCapacity
+                                        | BelowThreshold
+                                )
+                        {
+                            address == owner && **valid_sig
+                        } else if
+                        // Bonding and unbonding may affect validator sets
+                        matches!(
+                            pre,
+                            Consensus | BelowCapacity | BelowThreshold
+                        ) && matches!(
+                            post,
+                            Consensus | BelowCapacity | BelowThreshold
+                        ) {
+                            true
+                        } else {
+                            // Unknown state changes are not allowed
+                            false
+                        }
+                    }
+                    (None, Some(_post)) => {
+                        // Becoming a validator must be authorized
+                        **valid_sig
```

Comment:
maybe unnecessary, but should we have a `address == owner && *valid_sig` here too?